tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+                                        *state = Done;
+                                        continue 'outer;
+                                    }
+                                    Err(actual) => {
+                                        curr = actual;
+                                        continue;
+                                    }
+                                }
+                            }
+                            _ => unreachable!(),
+                        }
+
+                        // The state has been transitioned to waiting
+                        break;
+                    }
+
+                    // Safety: called while locked.
+                    unsafe {
+                        (*waiter.as_mut().get()).waker = Some(cx.waker().clone());
+
+                        // Insert the waiter into the linked list
+                        waiters.push_front(waiter.as_mut());
+                    }
+
+                    *state = Waiting;
+                }
+                Waiting => {
+                    let mut waiters = notify.waiters.lock().unwrap();
+
+                    {
+                        // Safety: called while locked
+                        let w = unsafe { &mut *waiter.as_mut().get() };
+
+                        if w.notified {
+                            w.waker = None;
+                            w.notified = false;
+
+                            // Remove the entry from the linked list. The entry
+                            // may only be in the linked list while the state is
+                            // `Waiting`.
+                            //
+                            // Safety: called while the lock is held
+                            unsafe {
+                                waiters.remove(waiter.as_mut());
+                            }
+
+                            *state = Done;
+                        } else {
+                            if !w.waker.as_ref().unwrap().will_wake(cx.waker()) {
+                                w.waker = Some(cx.waker().clone());
+                            }
+
+                            return Poll::Pending;
+                        }
+                    }
+                }
+                Done => {
+                    return Poll::Ready(());
+                }
+            }
+        }
+    }
+}
+
+impl Drop for RecvFuture<'_> {
+    fn drop(&mut self) {
+        use RecvState::*;
+
+        // Safety: The type only transitions to a "Waiting" state when pinned.
+        let (notify, state, mut waiter) = unsafe { Pin::new_unchecked(self).project() };
+
+        // This is where we ensure safety. The `RecvFuture` is being dropped,
+        // which means we must ensure that the waiter entry is no longer stored
+        // in the linked list.
+        if let Waiting = *state {
+            let mut notify_state = WAITING;
+            let mut waiters = notify.waiters.lock().unwrap();
```

Comment:
If the mutex is poisoned, this will double panic. But, if my reading of the comment above is correct, if we fail to remove the waiter from the list, we will have corrupted the list by leaving a pointer dangling. Should we instead "force" the mutex to unlock even if it is poisoned (e.g. by using the guard returned in a `PoisonError`), so that we can continue unwinding without leaving a dangling entry in the list?

Reply:
How would you expect the mutex to be able to become poisoned without being in a state that is so undefined that it doesn't matter anymore how we try to handle it?

Reply:
Hmm, you're right, I thought there were a couple cases where we made assertions while the lock is held, but after taking another look, that doesn't seem to be the case. Still, I'd prefer to err on the side of resilience if future changes add possible panics while holding a lock...